File: 프로그래머스/lv2/60057. 문자열 압축/문자열 압축.py
def solution(s):
    answer = len(s)
    for step in range(1, len(s)// + 1):
        sentence = ""
        prev = s[:step]
        count = 1
        for i in range(step, len(s), step):
            if s[i: i+step] == prev:
                count += 1
            else:
                sentence += str(count) + prev if count > 1 else prev
                prev = s[i: i+step]
                count = 1
        sentence += str(count) + prev if count > 1 else prev
        answer = min(answer, len(sentence))
    return answer

File: 프로그래머스/2/60057. 문자열 압축/문자열 압축.java
class Solution {
    public int solution(String s) {
        int n = s.length();
        if (n == 1) return 1;

        int answer = n;
        for (int unit = 1; unit <= n / 2; unit++) {
            StringBuilder sb = new StringBuilder();
            String prev = s.substring(0, unit);
            int count = 1;
            int i;

            for (i = unit; i + unit <= n; i += unit) {
                String cur = s.substring(i, i + unit);
                if (cur.equals(prev)) {
                    count++;
                } else {
                    if (count > 1) sb.append(count);
                    sb.append(prev);
                    prev = cur;
                    count = 1;
                }
            }

            if (count > 1) sb.append(count);
            sb.append(prev);

            if (i < n) sb.append(s.substring(i));

            answer = Math.min(answer, sb.length());
        }
        return answer;
    }
}
